File: frontend/src/features/notifications/renderNotification.ts
import type { NotificationPublic } from "@/client"

/**
 * Notification rows carry `event_type` + a structured `payload`, never
 * rendered text (see phase-9-in-app-notifications.md decision 6). This is
 * the one place that turns that structured data into copy, so wording
 * changes never require a migration.
 */
export function renderNotificationText(
  notification: NotificationPublic,
): string {
  const { event_type, payload } = notification

  if (event_type === "file_shared") {
    const sharerEmail = payload.sharer_email
    const fileName = payload.file_name
    if (typeof sharerEmail === "string" && typeof fileName === "string") {
      return `${sharerEmail} shared "${fileName}" with you`
    }
  }

  return "New notification"
}

File: mobile/lib/features/notifications/domain/render_notification.dart
import 'package:cloudestorage/features/notifications/domain/notification_models.dart';

/// Notifications carry `event_type` and a structured `payload`, never
/// rendered text (phase-9-in-app-notifications.md decision 6). This mirrors
/// `frontend/src/features/notifications/renderNotification.ts` by design --
/// keep wording in step across both clients, or the same event reads
/// differently on each platform.
String renderNotificationText(AppNotification notification) {
  if (notification.eventType == 'file_shared') {
    final sharerEmail = notification.payload['sharer_email'];
    final fileName = notification.payload['file_name'];
    if (sharerEmail is String && fileName is String) {
      return '$sharerEmail shared "$fileName" with you';
    }
  }

  return 'New notification';
}
